server/schedule/labeler/labeler.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2021 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package labeler
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule/rangelist"
+	"go.uber.org/zap"
+)
+
+// RegionLabeler is utility to label regions.
+type RegionLabeler struct {
+	storage *core.Storage
+	sync.RWMutex
+	labelRules map[string]*LabelRule
+	rangeList  rangelist.List // sorted LabelRules of the type `KeyRange`
+}
+
+// NewRegionLabeler creates a Labeler instance.
+func NewRegionLabeler(storage *core.Storage) (*RegionLabeler, error) {
+	l := &RegionLabeler{
+		storage:    storage,
+		labelRules: make(map[string]*LabelRule),
+	}
+
+	if err := l.loadRules(); err != nil {
+		return nil, err
+	}
+	return l, nil
+}
+
+func (l *RegionLabeler) loadRules() error {
+	var toDelete []string
+	err := l.storage.LoadRegionRules(func(k, v string) {
+		var r LabelRule
+		if err := json.Unmarshal([]byte(v), &r); err != nil {
+			log.Error("failed to unmarshal label rule value", zap.String("rule-key", k), zap.String("rule-value", v), errs.ZapError(errs.ErrLoadRule))
+			toDelete = append(toDelete, k)
+			return
+		}
+		if err := l.adjustRule(&r); err != nil {
+			log.Error("failed to adjust label rule", zap.String("rule-key", k), zap.String("rule-value", v), zap.Error(err))
+			toDelete = append(toDelete, k)
+			return
+		}
+		l.labelRules[r.ID] = &r
+	})
+	if err != nil {
+		return err
+	}
+	for _, d := range toDelete {
+		if err = l.storage.DeleteRegionRule(d); err != nil {
+			return err
```

Comment:
Should we continue deleting the rest of `toDelete` rather than returning here?

Reply:
If an error is encountered here, it is possible that the underlying etcd storage cannot be written to, and the next operation will most likely fail. I think it would make more sense to delay the deletion until the next load.